fenix/docs/architecture-overview.md

```diff
@@ -2,168 +2,177 @@
 
-#### Description
-Simple data object that carries information about a [State](#state) change to a [Store](#store).  An Action describes _something that happened_, and carries any data relevant to that change. For example, `AccountSettingsFragmentAction.SyncFailed(val time: Long)`, which describes a sync that failed at a specific time.
+Screen-based Stores should be created using [StoreProvider.get](https://github.com/mozilla-mobile/firefox-android/blob/main/fenix/app/src/main/java/org/mozilla/fenix/components/StoreProvider.kt).
 
 -------
 
-### <a name="state"/>State
-#### Overview
-Description of the state of a screen
-
-See [mozilla.components.lib.state.State](https://github.com/mozilla-mobile/android-components/blob/main/components/lib/state/src/main/java/mozilla/components/lib/state/State.kt)
+### <a name="state">State</a>
+#### **Overview**
+Description of the state of a screen or other area of the app.
 
-Referenced by: [Store](#store)
+See [mozilla.components.lib.state.State](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/lib/state/src/main/java/mozilla/components/lib/state/State.kt)
 
-#### Description
+#### **Description**
 Simple, immutable data object that contains all of the backing data required to display a screen. This does not include style details like colors and view sizes, which are handled by the [View](#view).
 
 As much as possible, the State object should be an accurate, 1:1 representation of what is actually shown on the screen. That is to say, the screen should look exactly the same any time a State with the same values is emitted, regardless of any previous changes. This is not always possible as Android UI elements are very stateful, but it is a good goal to aim for.
 
 One major benefit of rendering a screen based on a State object is its impact on testing. UI tests are notoriously difficult to build and maintain. If we are able to build a simple, reproducible [View](#view) (i.e., if we can trust that the View will render as expected), that allows us to test our UI by verifying the correctness of our State object.
 
-This also gives us a major advantage when debugging. If the UI looks wrong, check the State object. If it's correct, the problem is in the View binding. If not, check that the correct [Action](#action) was sent. If so, the problem is in the reducer. If not, check the [Controller](#controller) that sent the Action. This helps us quickly narrow down problems.
+This also gives us a major advantage when debugging. If the UI looks wrong, check the State object. If it's correct, the problem is in the View. If not, check that the correct [Action](#action) was sent. If so, the problem is in the reducer. If not, check the component that sent the Action. This helps us quickly narrow down problems.
```

Comment:
I want to be careful about being overly prescriptive here - I'm not sure that modeling the state that way will make sense in every case and I would worry that we might lead people down the wrong path. Maybe a reference to an `Additional Reading` section that contains it?

I would also potentially like to chat more about this idea offline. I gave it a shot while reworking the "simplified example" referenced in this document, which will be included in a follow-up patch. I thought I would update the example to use a basic history screen instead, since that's the first feature area I'll be refactoring following these docs updates. 

One issue I found with it is that it forces the reducers and middlewares to act as finite state machines. That is, both components need to do a lot of additional type checking/casting in order to ensure the previous state matches expectations. That, and some of the current UI state needs to be propagated through most subtypes. Two of the features of history include multiselect and deletion of items. During deletion, the items are kept in memory in case of an error at the storage layer. Below is a comparison of the two styles of modeling state if we following the existing behavior of history:

Both can use the same list of actions:
```
sealed class HistoryAction {
    object Init : HistoryAction()
    data class ItemsChanged(val items: List<History>) : HistoryAction()
    data class DeleteItems(val items: List<History>) : HistoryAction()
    data class DeleteFinished(val deletedItems: List<History>) : HistoryAction()
    data class ToggleItemSelection(val item: History) : HistoryAction()
}
```

Sealed types:
```
sealed class HistoryState {
    object Loading : HistoryState()
    data class Loaded(val items: List<History>) : HistoryState()
    data class Multiselecting(val items: List<History>, val selectedItems: List<History>) : HistoryState()
    data class Deleting(val items: List<History>, val itemsBeingDeleted: List<History>) : HistoryState()
    data class DeletingAndMultiselecting(val items: List<History>, val itemsBeingDeleted: List<History>, val selectedItems: List<History>) : HistoryState()
}

fun reducer(oldState: HistoryState, action: HistoryAction): HistoryState = when (action) {
    is HistoryAction.ToggleItemSelection -> when (oldState) {
        is HistoryState.Multiselecting -> {
            // add or remove item from selected and leave multiselect if necessary
        }

        is HistoryState.Loaded -> {
            // add item to selected and enter multiselecting
        }

        is HistoryState.Deleting -> {
            // add item to selected and enter multiselecting and deleting
        }

        is HistoryState.DeletingAndMultiselecting -> {
            // add or remove item from selected and leave multiselect if necessary 
        }

        else -> {
            // this is an illegal state, handle appropriately
        }
    }
    is HistoryAction.DeleteItems -> {
        is HistoryState.Multiselecting -> {
            // exit multiselecting and delete selected items
        }
        is HistoryState.DeletingAndMultiselecting -> {
            // exit multiselecting and delete selected items
        }
        else -> {
            // the rest are all probably illegal states and need to be handled appropriately
        }
    }
}
```

data class:
```
data class HistoryState(
    val items: List<History>,
    val selectedItems: List<History>,
    val itemsBeingDeleted: List<History>,
)

fun reducer(oldState: HistoryState, action: HistoryAction): HistoryState = when (action) {
    is HistoryAction.DeleteItems -> oldState.copy(itemsBeingDeleted = action.items)
    is HistoryAction.ToggleItemSelection -> {
        // the UI can know to whether display multiselect mode based on the selected items
        if (oldState.selectedItems.contains(action.item)) {
            oldState.copy(selectedItems = oldState.selectedItems - action.item)
        } else {
            oldState.copy(selectedItems = oldState.selectedItems + action.item)
        }
    }
}
```

The sealed class example is probably naive and could be improved, and I do ultimately think that direction will lead to a more correct program. That said, I think it will be slower to write and potentially harder to read.